synapse/storage/databases/main/user_directory.py

```diff
@@ -85,19 +85,17 @@ def _make_staging_area(txn):
             self.db_pool.simple_insert_many_txn(txn, TEMP_TABLE + "_rooms", rooms)
             del rooms
 
-            # If search all users is on, get all the users we want to add.
-            if self.hs.config.user_directory_search_all_users:
-                sql = (
-                    "CREATE TABLE IF NOT EXISTS "
-                    + TEMP_TABLE
-                    + "_users(user_id TEXT NOT NULL)"
-                )
-                txn.execute(sql)
+            sql = (
+                "CREATE TABLE IF NOT EXISTS "
+                + TEMP_TABLE
+                + "_users(user_id TEXT NOT NULL)"
+            )
+            txn.execute(sql)
 
-                txn.execute("SELECT name FROM users")
-                users = [{"user_id": x[0]} for x in txn.fetchall()]
+            txn.execute("SELECT name FROM users")
+            users = [{"user_id": x[0]} for x in txn.fetchall()]
 
-                self.db_pool.simple_insert_many_txn(txn, TEMP_TABLE + "_users", users)
+            self.db_pool.simple_insert_many_txn(txn, TEMP_TABLE + "_users", users)
```

Comment:
This isn't your fault, but since we're now going to start doing this for all servers: Is this going to be OK on big servers (like matrix.org?)?

Maybe we can use `INSERT INTO temp_users (user_id) SELECT name FROM users` to avoid having to pull it into Synapse only to shove it back out? (I checked this works on (recent) SQLite and Postgres. Hopefully we're not tripped up by older SQLites, but guess we could only do this for Postgres if there's a problem.)

I imagine the database would be smart enough to stream the data directly from one table to another (and save us buffering it into memory).

Reply:
I'm not sure how to answer that question without a big server to try it on. :( Worth noting that this is only at the point of rebuilding the directory, which is apparently a slow process anyway.

Another thought: we could `INSERT INTO user_directory` directly `FROM profiles`. We'd also need to `INSERT INTO user_directory_search`. And we'd need to consider what happens if someone adds to the profiles table in between the insertions. (Lock the profiles table?)

Reply:
> And we'd need to consider what happens if someone adds to the profiles table in between the insertions. (Lock the profiles table?)

Locking sounds slightly no-no-y (depending on how long we're talking).

But when someone updates their profile, it goes into the directory anyway (or will when you add that change in)! :D

> Another thought: we could INSERT INTO user_directory directly FROM profiles

I do like the sound of that. I don't know how much overhaul it'd want from you and whether you'd want to do it or not.
(Just be careful: I don't think it's wise to try and do that in one go, because with concurrent writers to the `user_directory` table, it might happen that you get a serialisation failure in Postgres and have to start the transaction all over again.)

Reply:
> I don't know how much overhaul it'd want from you and whether you'd want to do it or not.

Not against doing it, but I'm not chomping at the bit to get that started now.

> But when someone updates their profile, it goes into the directory anyway (or will when you add that change in)! :D

Oh yeah, that's right.

> it might happen that you get a serialisation failure in Postgres and have to start the transaction all over again.

How would that manifest? A certain kind of error from psycopg2 that we'd need to catch and retry within the application? Do we have to handle that generally across synapse?

> with concurrent writers to the user_directory table, 

Erm, don't we already have concurrent writers to the `user_directory` table? I think the things competing for writes are:

- all of the stuff in the `UserDirectoryHandler`, which only ever runs on one worker
   - there's a scary looking comment in its docstring: `N.B.: ASSUMES IT IS THE ONLY THING THAT MODIFIES THE USER DIRECTORY`
- the background process in the `UserDirectoryBackgroundUpdateStore`
- anything that calls `UserDirectoryHandler.handle_local_profile_change`. (Register, Activate, Deactivate, Update Profile).

Reply:
> How would that manifest? A certain kind of error from psycopg2 that we'd need to catch and retry within the application? Do we have to handle that generally across synapse?

Yup! `runInteraction` (or one of the things it calls) will automatically retry those up to 5 times (iirc), so your transaction might actually get run multiple times!

> Erm, don't we already have concurrent writers to the user_directory table?

Yes, and it's not too much hassle if they get restarted, because they're small transactions that only rarely collide. But if you have a single transaction writing (and potentially colliding with any one of) 100k+ rows, that's a *big* transaction to have to restart (and if it's slow enough and collision-y enough, might not succeed within 5 retries).

NB this is my understanding; I don't know all the subtleties about Postgres.
The relevant section of the manual is (I think we use this level) [REPEATABLE READ transaction isolation](https://www.postgresql.org/docs/13/transaction-iso.html#XACT-REPEATABLE-READ).

Reply:
> (I think we use this level) REPEATABLE READ transaction isolation.

That's correct, for the record. It gets set at https://github.com/matrix-org/synapse/blob/v1.42.0/synapse/storage/engines/postgres.py#L107-L109.

On the topic of this thread: slurping all the users into synapse just to spit them out again does seem *particularly* stupid and replacing it with a `INSERT INTO temp_users (user_id) SELECT name FROM users` might be a quick win.  And yes, no doubt you could go further in optimising this code too.

Still, I'd say that any such optimisations are orthogonal to this PR, and my inclination would be to leave it as is. Particularly given this code only runs when the directory is first populated (at which point it is reasonable to expect a small `users` table), I don't think we need to worry too much.

Reply:
@richvdh I think we will wind up wanting to do a rebuild once all of David's work lands, to make sure that the user directory is scrubbed of private profiles. In that case, the fact that big servers such as `matrix.org` now need to hit this code, I think it would be worthwhile doing at least that.
(I agree that this could easily be a tiny separate PR, though.)

Reply:
I'll make a note to revisit this in a future PR.